File: AlgorithmInCSharp/JobInterviewPrepare/JobInterviewPrepare/Data Strcuture/DoublyLL.cs

using JobInterviewPrepare;
namespace JobInterviewPrepare.Data_Strcuture
{
    public class DoublyLL
    {
        Node head;
        Node tail;
        int value;
        public DoublyLL()
        {
            value = 0;
            head = null;
            tail = null;
        }
        public DoublyLL(int value)
        {
            this.value = value;
            head = null;
            tail = null;
        }
        public void InsertFront(DoublyLL doublyLL, int value)
        {
            Node newNode = new Node(value);
            newNode.next = doublyLL.head;
            //newNode.prev = null;
            if(doublyLL.head != null)
            {
                //doublyLL.head.prev = newNode;
            }
            doublyLL.head = newNode;
        }
        // if the linked list is empty, then we simply add the new node as the Head of the LL.
        
        public void InsertLast(DoublyLL doublyLL, int value)
        {
        }
        public void GetLastNode(DoublyLL dll)
        {
        }
        // insert a new node after a give node.
        public void InsertAfterNode(Node _node, int value)
        {

            return;
        }

    }
}

// Internal vs Public method. 
// Algorithm  comparison.... where I can test it?  
// Managed code is a code that is executed by CLR(Command Language Runtime) 
// Net is considered as managed beceause of the net framework which internally uses the garbage collector to clear up the unused memory . (Important@!!!!!)


// Insert a node in a DDL .
// 1. isnert a new node before head.
// 2. Insert a new node at the end of the list
// 3. Isnert a new node in the middle of the list.
 

File: udacity/DS/P1/DoublyLL.py
class Node:
    def __init__(self, value):
        self.next = None
        self.prev = None
        self.value = value


class DoublyLL:

    def __init__(self):
        self.head = None
        self.tail = None

    def push(self, value):
        newNode = Node(value)
        if self.head is None:
            self.head = newNode
            self.tail = newNode
        else:
            newNode.next = self.head
            self.head.prev = newNode
            self.head = newNode

    def insertAfterNode(self, selectNode, value):
        if selectNode is None:
            return
        newNode = Node(value)
        newNode.next = selectNode.next 

        selectNode.next = newNode
        newNode.prev = selectNode
        if newNode.next is not None:
            newNode.next.prev = newNode


    def append(self, value):
        newNode = Node(value)
        if head is None:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next=  newNode
            newNode.prev = self.tail
            self.tail = newNode

    def traverse(self, node):
        while node != None:
            print(node.value)
            node = node.next



dd = DoublyLL()
dd.push(10)
dd.push(20)
dd.push(30)
dd.push(40)
dd.push(50)
dd.insertAfterNode(dd.head.next, 35)            
dd.traverse(dd.head)